src/Components/Artsy2.tsx

```diff
@@ -0,0 +1,66 @@
+import { ContextProps } from "Components/Artsy"
+import React, { SFC } from "react"
+import { getUser } from "Utils/getUser"
+import { createEnvironment } from "../Relay/createEnvironment"
+
+const Context = React.createContext<ContextProps>({})
+
+/**
+ * Creates a new Context.Provider with a user and Relay environment, or defaults
+ * if not passed in as props.
+ */
+export const ContextProvider: SFC<ContextProps> = ({ children, ...props }) => {
+  const currentUser = getUser(props.currentUser)
+  const relayEnvironment =
+    props.relayEnvironment || createEnvironment({ user: currentUser })
+
+  const providerValues = {
+    ...props,
+    _isNestedInProvider: true,
+    currentUser,
+    relayEnvironment,
+  }
+
+  return <Context.Provider value={providerValues}>{children}</Context.Provider>
+}
+
+interface ArtsyConsumerProps {
+  children: (ContextProps) => JSX.Element
+}
+
+/**
+ * Creates a new Context.Consumer and validates that a Context.Provider has
+ * been created.
+ */
+export const ContextConsumer: SFC<ArtsyConsumerProps> = ({ children }) => {
+  return (
+    <Context.Consumer>
+      {contextValues => {
+        // FIXME: Find way to exclude _isNested from check
+        if (!(contextValues as any)._isNestedInProvider) {
```

Comment:
@alloy - re typing, specifically this prop. I noticed in the other `Artsy` component that there's a [validation map](https://github.com/damassi/reaction/blob/02208759b96be69311e6ac3934c69cefb92532f5/src/Components/Artsy.tsx#L44-L45) that is implemented on the Class but not too sure how that would work here (if indeed that's what's allowing this prop to pass transparently through).

Reply:
I'll validate that a React warning is thrown and if so will do 👍